lib/apiservers/engine/backends/container.go

```diff
@@ -1605,6 +1599,70 @@ func (c *Container) containerAttach(name string, ca *backend.ContainerAttachConf
 	return nil
 }
 
+// ContainerRename changes the name of a container, using the oldName
+// to find the container. An error is returned if newName is already
+// reserved.
+func (c *Container) ContainerRename(oldName, newName string) error {
+	defer trace.End(trace.Begin(newName))
+
+	if oldName == "" || newName == "" {
+		err := fmt.Errorf("neither old nor new names may be empty")
+		log.Errorf("%s", err.Error())
+		return err
+	}
+
+	if !utils.RestrictedNamePattern.MatchString(newName) {
+		err := fmt.Errorf("invalid container name (%s), only %s are allowed", newName, utils.RestrictedNameChars)
+		log.Errorf("%s", err.Error())
+		return err
+	}
+
+	// Look up the container name in the metadata cache to get long ID
+	vc := cache.ContainerCache().GetContainer(oldName)
+	if vc == nil {
+		log.Errorf("Container %s not found", oldName)
+		return NotFoundError(oldName)
+	}
+
+	oldName = vc.Name
+	if oldName == newName {
+		err := fmt.Errorf("renaming a container with the same name as its current name")
+		log.Errorf("%s", err.Error())
+		return err
+	}
+
+	if exists := cache.ContainerCache().GetContainer(newName); exists != nil {
+		err := fmt.Errorf("conflict. The name %q is already in use by container %s. You have to remove (or rename) that container to be able to re use that name.", newName, exists.ContainerID)
+		log.Errorf("%s", err.Error())
+		return derr.NewRequestConflictError(err)
+	}
+
+	// reserve the new name in containerCache
+	if err := cache.ContainerCache().ReserveName(vc, newName); err != nil {
+		log.Errorf("%s", err.Error())
+		return fmt.Errorf("Error when allocating new name: %v", err)
```

Comment:
This will only happen when a container named `newName` already exists - I think we should return a `derr.NewRequestConflictError` similar to the one above.